### PR TITLE
Enable degree heuristic for template matching.

### PR DESCRIPTION
**Pull request by abrodersen**
If one template is a subset of the other, we need a heuristic for determining which template is the match.

For example: 
````
{"foo":1,"bar":2}
````
should have a higher priority than
````
{"foo": 1}
````

A degree function has already been defined in the code. It's definition looks like:
````
degree (o: object) = o.keys().map(degree).sum()
degree (a: array) = a.map(degree).sum()
degree (_) = 1
````

### CODE REVIEW

**Comment by lurch** [comment]
@majorz @hedss I see in https://github.com/resin-io/reconfix/pull/40/commits/d58e05faa98588a15d589c02c723eed102025499 that @abrodersen is bumping the versions (all to the same semver-string) in each of `Cargo.toml`, `bindings/node/native/Cargo.toml` and `package.json`.
Bumping multiple files like this currently isn't supported by `versionist` - setting the [`updateVersion`](updateVersion) to `'npm'` would bump only `package.json`, setting the `updateVersion` to `'cargo'` would only bump `Cargo.toml`. So using [versionist](https://github.com/resin-io/versionist) (and also by implication VersionBot #26 ) means **a)** there's no way to bump `bindings/node/native/Cargo.toml`, and **b)** there's no way to bump multiple files at a time.

To solve **a)** I guess we could adopt the approach used by @nghiant2710 in https://github.com/resin-io/versionist/pull/100 and add an optional `targetFile` option to each of the presets? Or perhaps in the `npm` / `cargo` case a `targetDir` would be a better option (since the actual filenames are fixed)? This means you could specify e.g. `updateVersion: {preset: 'cargo', targetDir: 'bindings/node/native'}` in `versionist.conf.js` and then versionist would update `bindings/node/native/Cargo.toml` (and `bindings/node/native/Cargo.lock` too, if it existed). 

To solve **b)** would it be possible (I'm not sure how fine-grained JavaScript's type inspection is) to allow `updateVersion` to also be passed an array of values? So that you could specify the following in `versionist.conf.js`:
```
updateVersion: [
    'cargo',
    {
        preset: 'cargo',
        targetDir: 'bindings/node/native'
    },
   'npm'
]
```
and then versionist would update all three of the files  `Cargo.toml`, `bindings/node/native/Cargo.toml` and `package.json` .
Which I guess means the parameter-type for `updateVersion` then becomes something like "function or string or object or array of (functions and/or strings or and/objects)" :nerd_face: 

@abrodersen Do you think it'll be a common scenario to want to have versionist updating multiple separate files all to the same version, or am I simply creating a mountain out of a molehill? :wink: 
If people agree that this makes sense, and I'm not simply going mad, then I'll create the appropriate issues in the versionist repo.

EDIT: While writing the above comment, Aaron [also bumped](https://github.com/resin-io/reconfix/pull/40/commits/ef50032006014a527eff9fb54783d5b6aff3f108) the version in `package-lock.json` - and there's already a versionist issue for that :) https://github.com/resin-io/versionist/issues/103

**Comment by abrodersen** [comment]
@lurch I think it makes sense to keep the versions in sync, since the `bindings` and the underlying library are all kept in the same repository. While we could, in theory, make changes to `reconfix` without bumping the `bindings`, I think it would violate assumptions to change the behavior for a user of the bindings without bumping the version number.

I'm also not very familiar with `versionist`. What is needed to include it in this project?

**Comment by lurch** [comment]
> I'm also not very familiar with versionist. What is needed to include it in this project?

`versionist` is the underlying technology used by VersionBot, and you can read more about it at https://github.com/resin-io/process/blob/master/process/Commit_and_PR_Guidelines.md and https://github.com/resin-io/versionist

**Comment by lurch** [comment]
Pinging @lekkas too as I guess multiple-versioned-files in a repo probably counts as a wider process issue.

**Comment by majorz** [comment]
@lurch @hedss @lekkas although having a few Rust packages in the same repo is seen occasionally (maybe it is the same with npm packages), I have a feeling that it may not be a good idea for projects using versionist.

A single commit will target all or less packages. Since it may not target all of them, one option would be putting an extra footer field in the commit message with description of targeted packages - this will be very error prone.

Another option I could think of is scanning which files have changed for determining which packages need version update. This could be a better solution.

The bigger issue I see is that if a single commit targets multiple packages, what the changelog entry will be for each of those packages? One and the same for each of them does not feel right. That leads me thinking that at the bottom of it is a tighter-coupling kind of problem. The solution maybe: a separate repo for a single package, and multiple packages in a single repo a bad practice?

**Comment by majorz** [comment]
Actually now thinking further, if there is just a single CHANGELOG file in the root of the repo itself and also given that the version will be always one and the same across all packages, then extending `updateVersion` in the way @lurch suggests does make sense to me. I would go with exactly the same approach as well. The `lodash` library used in versionist provides the introspection capabilities for **b)**.

Updating the version number of an inner package without changing its content is not the cleanest approach for me, but if this fits the requirements of a repo, why not.

**Comment by lurch** [comment]
@abrodersen Do the bindings specify which version of the underlying library they target? If not, should they? Or is this where a `Cargo.lock` file would come in useful?
(I know almost nothing about Rust or Cargo)

**Comment by abrodersen** [comment]
@lurch Yes, the `Cargo.lock` file specifies the exact version of a dependency. So in the case of the Node.js bindings, we refer to the dependency by a file path (`../../..`) and the `Cargo.lock` file encodes the precise version of the dependency. However, it is common practice with Rust libraries to not include the `Cargo.lock` file in version control for libraries.

**Comment by lurch** [comment]
@majorz obviously knows more about this than me, but if we're bumping the version of the bindings every time we bump the version of the underlying library, then maybe it _does_ make sense to add the `Cargo.lock` for the bindings, so that if somebody installs versionX of the bindings, they know that it'll also be installing the matching versionX of the underlying library? (IYSWIM?) Or maybe there's some other way of doing that?

Or maybe I'm just talking out of my a\*\*e ? :rofl:

**Comment by lekkas** [comment]
@lurch 

 On bumping `bindings/node/native/Cargo.toml`:

> Or perhaps in the npm / cargo case a targetDir would be a better option 

IMO this feels feels like best option for me as well.

On supporting version bumping of multiple files at a time:

>  [...] the parameter-type for updateVersion then becomes something like "function or string or object or array of (functions and/or strings or and/objects)"

Adding more flexibility to the `updateVersion` function also sounds like a good idea and the example also makes sense to me from an API perspective. We just need to make sure that the semantics of the different parameter types are clearly documented in the `updateVersion` function implementation.

On the other issue that is mentioned in the thread

> multiple-versioned-files in a repo probably counts as a wider process issue.

As long as the version is aligned in all the separate places, it shouldn't be an issue. After all, we already use this process convention (i.e. aligned versions) in core versioning processes like `npm`, where `package.json`, `CHANGELOG.md`, commit message and commit tag are all aligned. I remember this point being raised before (iirc in a `versionist` issue thread) where we went with the keep-them-in-sync approach.

@lurch are you up to pursuing these ideas on Versionist? cc @hedss any objections here?

@abrodersen unless it's already been discussed, I think it might worth starting a thread in Flowdock, bring Juanchi and Alex in the discussion and decide on a way forward for the versioning and release lifecycle of the separate modules/libraries (i.e. Rust `node-reconfix`, `reconfix` and Node `reconfix` right?) that are hosted on this repo.

**Comment by lurch** [comment]
@lekkas Created three new issues on the versionist repo, which I _think_ covers everything?

**Comment by lekkas** [comment]
yeap, thanks @lurch!

File: src/transform.rs [hunk 1]
```diff
@@ -78,7 +78,12 @@ fn sort_files<'a>(files: &'a BTreeMap<String, File>) -> Vec<(&'a str, &'a File)>
     ordered
 }
 
+/// Given the wet JSON state and a list of mappings, returns the matched value
+/// or `None` if no values matched.
 fn extract_value(wet: &Value, mappings: &[Mapping]) -> Result<Option<Value>> {
+    // Generate a list of tuples containing the candidates for matching. The first
+    // tupel value is the degree of the template or `None` for direct mappings.
```
Comment (lurch) [inline src/transform.rs]: tupel -> tuple :wink:

**Pull request merged**